prolog.cl

```diff
@@ -841,6 +842,747 @@
+           "Revoke grant decisions from the user on the table."
+           "Revoke grant decisions from the user on the schema."
+           )
+)
+
+  (multi "A Data Engineer is implementing a near real-time ingestion pipeline to load data into Snowflake using the Snowflake Kafka connector. There will be three Kafka topics created.
+Which Snowflake objects are created automatically when the Kafka connector starts?"
+         4
+        '("Tables"
+          "Pipes"
+          "Internal stages")
+        
+         '("Tasks"
+           "External stages"
+           "Materialized views"
+           )
+)
+
+  (multi "What kind of Snowflake integration is required when defining an external function in Snowflake?"
+         4
+        '(API integration"")
+        
+         '("HTTP integration"
+           "Notification integration"
+           "Security integration"
+           )
+)
+
+  (multi "What is the purpose of the BUILD_STAGE_FILE_URL function in Snowflake?"
+         4
+        '("It generates a permanent URL for accessing files in a stage.")
+        
+         '("It generates an encrypted URL for accessing a file in a stage."
+           "It generates a staged URL for accessing a file in a stage."
+           "It generates a temporary URL for accessing a file in a stage."
+           )
+)
+
+  (multi "A Data Engineer needs to load JSON output from some software into Snowflake using Snowpipe.
+Which recommendation(s) apply to this scenario?"
+         4
+        '("Ensure that data files are 100-250 MB (or larger) in size, compressed."
+         "Extract semi-structured data elements containing null values into relational columns before loading."
+         "Verify each value of each unique element stores a single native data type (string or number).")
+        
+         '("Load large files (1 GB or larger)."
+           "Load a single huge array containing multiple records into a single table row."
+           "Create data files that are less than 100 MB and stage them in cloud storage at a sequence greater than once each minute."
+           )
+)
+
+  (multi "Which Snowflake feature facilitates access to external API services such as geocoders, data transformation, machine learning models, and other custom code?"
+         4
+        '("External functions")
+        
+         '("Security integration"
+           "External tables"
+           "Java User-Defined Functions (UDFs)"
+           )
+)
+
+(fact "For external functions. the remote service"
+      '("must" "doesn't have to")
+      "return one row for each row received."
+      )
+
+(fact "RETURNS NULL ON NULL INPUT"
+      '("will not" "will")
+      "call the User-Defined Function (UDF) if any input is null."
+      )
+
+
+  (multi "A Data Engineer is working on a Snowflake deployment in AWS eu-west-1 (Ireland). The Engineer is planning to load data from staged files into target tables using the COPY INTO command.
+Which sources are valid? (Choose 3)"
+         6
```

Comment:
Remove the 'Choose 3" instruction, we may in the future allow the examiner to pick fewer than all supplied choices, at which point it may not pick all three of the correct answer (i.e. it may pick 1 correct and 2 incorrect, or 2 each).

Also, if we limit to <6 choices, you get slightly different question composition on subsequent runs through the exam, which I find helps with retention.